src/tokenserver/error.rs

```diff
@@ -0,0 +1,185 @@
+            name: "",
+            description,
+            http_status: StatusCode::UNAUTHORIZED,
+        }
+    }
+
+    pub fn invalid_client_state(description: &'static str) -> Self {
+        Self {
+            status: "invalid-client-state",
+            location: ErrorLocation::Header,
+            name: "",
+            description,
+            http_status: StatusCode::UNAUTHORIZED,
+        }
+    }
+
+    pub fn internal_error() -> Self {
+        Self {
+            status: "internal-error",
+            location: ErrorLocation::Internal,
+            name: "",
+            description: "Server error",
+            http_status: StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
+
+    pub fn unsupported(description: &'static str) -> Self {
+        Self {
+            status: "error",
+            location: ErrorLocation::Url,
+            name: "",
+            description,
+            http_status: StatusCode::NOT_FOUND,
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum ErrorLocation {
+    Header,
+    Url,
+    Body,
+    Internal,
+}
+
+impl fmt::Display for ErrorLocation {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Header => write!(f, "header"),
+            Self::Url => write!(f, "url"),
+            Self::Body => write!(f, "body"),
+            Self::Internal => write!(f, "internal"),
+        }
+    }
+}
+
+impl fmt::Display for TokenserverError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            serde_json::to_string(&self).map_err(|_| fmt::Error)?
+        )
+    }
+}
+
+impl ResponseError for TokenserverError {
+    fn error_response(&self) -> HttpResponse {
+        HttpResponse::build(self.http_status).json(ErrorResponse::from(self))
+    }
+
+    fn status_code(&self) -> StatusCode {
+        self.http_status
+    }
+}
+
+struct ErrorResponse {
+    status: &'static str,
+    errors: [ErrorInstance; 1],
+}
+
+struct ErrorInstance {
+    location: ErrorLocation,
+    name: &'static str,
+    description: &'static str,
+}
+
+impl From<&TokenserverError> for ErrorResponse {
+    fn from(error: &TokenserverError) -> Self {
+        ErrorResponse {
+            status: error.status,
+            errors: [ErrorInstance {
+                location: error.location,
+                name: error.name,
+                description: error.description,
+            }],
+        }
+    }
+}
+
+impl Serialize for ErrorInstance {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(3))?;
+        map.serialize_entry("location", &self.location.to_string())?;
+        map.serialize_entry("name", &self.name)?;
+        map.serialize_entry("description", &self.description)?;
+        map.end()
+    }
+}
+
+impl Serialize for ErrorResponse {
```

Comment:
The error format on the new Tokenserver should be exactly the same as that on the legacy Tokenserver